pjmedia/src/pjmedia/sdp_neg.c

```diff
@@ -1793,3 +1946,167 @@ PJ_DEF(pj_status_t) pjmedia_sdp_neg_fmt_match(pj_pool_t *pool,
+    if (pj_strncmp(codec, &telephone_event, telephone_event.slen) == 0)
+        pt_check = PJMEDIA_RTP_PT_TELEPHONE_EVENTS;
+    else
+        pt_check = START_DYNAMIC_PT;
+    
+    for (; i < used_pt_num && i < max_pt_cnt; ++i) {
+        if (pj_strcmp(codec, &used_pt->map[i].codec) == 0) {
+            return used_pt->map[i].pt;
+        }
+    }
+    i = 0;
+
+    while ((i < used_pt_num) && (i < max_pt_cnt) && (pt_check <= 127))
+    {
+        if (pt_check == used_pt->map[i].pt) {
+            pt_check++;
+            i = 0;
+        } else {
+            i++;
+        }
+    }
+    if (pt_check > 127 || i == max_pt_cnt) {
+        /* No more available PT */
+        return 0;
+    }
+
+    used_pt->map[used_pt->map_cnt].pt = pt_check;
+    pj_strdup(pool, &used_pt->map[used_pt->map_cnt++].codec, codec);
+
+    return pt_check;
+}
+
+PJ_DEF(pj_status_t) pjmedia_sdp_neg_validate_pt(pjmedia_sdp_neg* neg,
+                                                pjmedia_sdp_session* sdp_sess)
+{
+#if PJMEDIA_SDP_NEG_MAINTAIN_SESSION_PT
+    pj_status_t status = PJ_SUCCESS;
+    unsigned i = 0;
+
+    for (; i < sdp_sess->media_count && i < neg->med_pt_map_cnt; ++i) {
+        unsigned j = 0;
+        unsigned attr_cnt = 0;
+        pjmedia_sdp_attr* new_attr_list[PJMEDIA_MAX_SDP_ATTR];
+        pjmedia_sdp_attr* ori_attr_list[PJMEDIA_MAX_SDP_ATTR];
+        pj_str_t ori_fmt[PJMEDIA_MAX_SDP_FMT];
+        pjmedia_sdp_media* sess_media = sdp_sess->media[i];
+        media_pt_map used_pt_map;
+        media_pt_map* med_pt_map = get_media_pt_map(neg, &sdp_sess->media[i]->desc.media);
+
+        if (!med_pt_map) {
+            PJ_LOG(4, (THIS_FILE, "Unable to find media pt map for [%.*]",
+                       sdp_sess->media[i]->desc.media.slen, 
+                       sdp_sess->media[i]->desc.media.ptr));
+            continue;
+        }
+
+        pj_memcpy(&used_pt_map, med_pt_map, sizeof(used_pt_map));
+        pj_memcpy(&ori_fmt, sess_media->desc.fmt, sizeof(ori_fmt));
+
+        for (; j < sess_media->desc.fmt_count; ++j) {
+            pjmedia_sdp_attr* attr = NULL;
+            pj_str_t pt_str = sess_media->desc.fmt[j];
+            unsigned pt = pj_strtoul(&pt_str);
+
+            if (pj_strtoul(&sess_media->desc.fmt[j]) < START_DYNAMIC_PT)
+                continue;
+
+            attr = pjmedia_sdp_media_find_attr2(sess_media, "rtpmap", &pt_str);
+            if (attr) {
+                pj_str_t codec;
+                unsigned p_exists;
+                unsigned pt_len = sess_media->desc.fmt[j].slen + 1;
+
+                codec.ptr = attr->value.ptr + pt_len;
+                codec.slen = attr->value.slen - pt_len;
+
+                ori_attr_list[attr_cnt] = attr;
+
+                p_exists = pt_exists(&used_pt_map, pt, &codec);
+                if (p_exists == 1 || p_exists == 2) {
+                    /* pt is assigned to another codec, find existing pt used by the
+                     * codec or a new one.
+                     */
```

Comment:
There are two cases in the check, but only one is described in the comment, better put description for both cases.